Persistence testkit implementation #15571

manonthegithub opened the pull request:
<!--
# Pull Request Checklist

* [ ] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [ ] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you updated the documentation?
* [ ] Have you added tests for any changed functionality?
-->
## Purpose
The implementation of persistence testkit module.
<!-- What does this PR do? -->

## References

<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please DON`T use `Fixes` notation.
-->
References #15571

## Changes
I have finally come to the point when it is almost ready. I have added Java api, documentation, improved scala api, added many more tests.
<!-- Bullets for important changes in this PR -->

## Background Context
Now I need more people to look at it.

Looking forward for your comments!
<!-- Why did you take this approach? -->

manonthegithub committed:
Persistence testkit implementation #15571

* Testkit for persistent actors
* Documentation for the testkit

manonthegithub committed:
Fix Java compatibility issue

* Took the JournalOperations out of MessageStorage object
* Took the SnapshotOperations out of SnapshotStorage object

manonthegithub committed:
Fix scala 2.13 compatibility issue

raboof commented:
OK TO TEST

raboof commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -1105,6 +1105,170 @@ For more advanced schema evolution techniques refer to the @ref:[Persistence - S
 
 ## Testing
 
+### Persistence TestKit
```
I wonder if 'testing' now deserves its own page

manonthegithub replied:
agree, I was planning to add the examples for Akka Typed.

manonthegithub committed:
formatting

manonthegithub committed:
fix for deprecated method use

manonthegithub committed:
fix for config docs

manonthegithub commented:
Rebased on the latest master and made fixes

manonthegithub committed:
replaced arrow symbol with => for scala 2.13

manonthegithub committed:
replaced arrow symbol with <- for scala 2.13

manonthegithub committed:
more fixes for scala 2.13

manonthegithub committed:
docs fix

manonthegithub commented:
Finally, made it work. I think this can be a release candidate. Do not plan to make any changes or additions until getting comments from you. So feel free to comment :).

patriknw commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -819,6 +819,170 @@ For more advanced schema evolution techniques refer to the @ref:[Persistence - S
+There are two testkit classes which have similar api:
+
+ * @scala[`PersistenceTestKit` class]@java[`PersistenceTestKit` class] is for messages
+ * @scala[`SnapshotTestKit` class]@java[`SnapshotTestKit` class] is for snapshots
```
since same names, no need to use the `@scala` and `@java`

manonthegithub replied:
corrected

patriknw commented on akka-persistence-testkit/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,30 @@
+    # enable serialization of the persisted messages
+    serialize = true
+    # timeout for assertions
+    assert-timeout = 1s
```
we use `3s` for `single-expect-default` so let's use that here also

manonthegithub replied:
corrected

patriknw reviewed (commented):
I had a quick look and it seems promising. I think it would be useful to have this.
As far as I can see it's not specific to Classic or Typed so we can probably have same testkit for both.

manonthegithub committed:
small docs fix and config fix

manonthegithub commented:
> I had a quick look and it seems promising. I think it would be useful to have this.
> As far as I can see it's not specific to Classic or Typed so we can probably have same testkit for both.

I think this should work. There is the only dependency on ActorSystem and on the journals. If the old journals are used by Typed persistence it should work.

renatocaval commented:
I'm interested to know the scope of this. I will have a look in detail soon, but I have a few questions already.

* do we want to have one testkit that covers classic and typed API? 
* does the testkit will allow me to test which messages were persisted or only the state of the actor?
* is it possible to add a check to verify that state, events and commands have a proper serializer? 

That last can be the object of another PR of course. The first two will probably be impacted by the design we choose right now.

manonthegithub commented:
> I'm interested to know the scope of this. I will have a look in detail soon, but I have a few questions already.
> 
> do we want to have one testkit that covers classic and typed API?

I did not have time to look into the Typed api. If it uses the same persistence api under the hood then the testkit should be compatible with it too, i.e. universal.
> does the testkit will allow me to test which messages were persisted or only the state of the actor?

It allows you to test the exact messages and also the order.

> is it possible to add a check to verify that state, events and commands have a proper serializer?

You can enable serialization, and see if the message have been serialized without an error.

> That last can be the object of another PR of course. The first two will probably be impacted by the design we choose right now.

You may look into the tests for a better understanding of the api. In the tests you have all the examples of the usage of the api.

raboof reviewed (commented):
Since this is a testkit module we're free to change the API later, so this is relatively low-risk

Some relatively superficial reviewing here - generally I like it, but I have to take some more time to understand the implications of the 'policy' approach better later

manonthegithub commented on akka-persistence-testkit/src/test/resources/application.conf:
```diff
@@ -0,0 +1 @@
+akka.actor.allow-java-serialization = true
```
We are serializing custom case class in the tests, for simplicity this class is serialized using standard java-serializer. I can also write a custom serializer and configuration for it.

manonthegithub commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -819,6 +819,170 @@ For more advanced schema evolution techniques refer to the @ref:[Persistence - S
+
+There are two testkit classes which have similar api:
+
+ * `PersistenceTestKit` class is for messages
```
@unidoc did not work, maybe you meant @apidoc?

raboof replied:
ah, sorry, yes that's what I meant

manonthegithub committed:
minor fixes

manonthegithub committed:
rebase onto latest master and test examples formatting

manonthegithub committed:
use akka converters impl instead of deprecated scala

manonthegithub committed:
scalafmt

manonthegithub commented:
rebased on the latest master. made minor improvements.

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/scaladsl/PersistenceTestKit.scala:
```diff
@@ -0,0 +1,507 @@
+
+}
+
+object SnapshotTestKit {
```
we don't guarantee binary compatibility for testkits so we have some freedom to change, but could be good to highlight that this is a new module that is more likely to change by adding the `@ApiMayChange` annotation at a few strategic places

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala:
```diff
@@ -0,0 +1,604 @@
+  private val parallelism = Runtime.getRuntime.availableProcessors()
+
+  private final val eventsMap: ConcurrentHashMap[K, Vector[InternalRepr]] =
+    new ConcurrentHashMap()
```
Given that this is a testkit without high performance requirements it would be possible to have a completely `synchronized` implementation and using ordinary Scala collections, but we can start like this and consider changing that if it becomes difficult to maintain or add more features.

patriknw reviewed (commented):
I think this is looking good. I have some opinions about naming of "message" vs "event".

patriknw commented:
Necessary extension of this would be to support queries, but that should be added separately.

manonthegithub commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -806,6 +806,170 @@ For more advanced schema evolution techniques refer to the @ref:[Persistence - S
+ 
+The testkit classes have two corresponding plugins which emulate the behavior of the storages: 
+
+ * @apidoc[PersistenceTestKitPlugin] class emulates a messages storage 
```
ok.

manonthegithub commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -806,6 +806,170 @@ For more advanced schema evolution techniques refer to the @ref:[Persistence - S
+ * @scala[`def clearByIdPreservingSeqNumbers(persistenceId: String): Unit`]@java[`public void clearByIdPreservingSeqNumbers(String persistenceId)`]
+Delete all the messages from the storage by persistence id and keep highest sequence number for it. 
+ * @scala[`def withPolicy(policy: Policies.PolicyType): this.type`]@java[`public PersistenceTestKit withPolicy(ProcessingPolicy<JournalOperation>) policy)`]
+You can set your own [policy](#setting-your-own-policy-for-the-storage) which emulates work of a snapshot storage. 
```
good catch, I'll check that

manonthegithub commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala:
```diff
@@ -0,0 +1,604 @@
+  private final val eventsMap: ConcurrentHashMap[K, Vector[InternalRepr]] =
+    new ConcurrentHashMap()
+
+  def forEachKey(f: K => Unit): Unit = eventsMap.forEachKey(parallelism, f)
```
agree :)

manonthegithub commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala:
```diff
@@ -0,0 +1,604 @@
+
+  /**
+   *
+   * Note! `elems` is call by name to preserve atomicity
```
> it's ok to have some `@InternalApi` in the other public files also, but when it's all internal it's good to make it clear by placing it in `internal` package

@patriknw it is not 100% internal. I have put the `JournalOperation` and `SnapshotOperation` and their impls. also into this file. These messages are used in the policies to react on the type of operation performed on the storage. I don't think these messages should be in the internal package, because they can be used by external developers. 
Should I put them in a different file, or just leave `InMemStorage` in the testkit package?

manonthegithub replied:
> what does atomicity mean here? why is it needed if elems is immutable?

Maybe not the best comment, I'll change. I use mapAny to obtain the elems in the implemenation:
https://github.com/akka/akka/blob/ee2d55ba04f987240d12bd4a908b4ce3759059de/akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala#L226-L227

This mapAny should be invoked inside the `compute` method  of `eventsMap` to preserve the atomicity of mutations on the maps of `seqNumbers` and `eventsMap`.
I think I will rework this part, to make it clearer.

manonthegithub committed:
improvemnts:
* improved internal storage impl
* renamed messages to events
* added docs for akka typed
* moved some classes to internal package
* added ApiMayChange annotations to some classes

manonthegithub committed:
added a few tests for typed actors

manonthegithub committed:
formatting

manonthegithub committed:
headers

manonthegithub committed:
formatting docs

manonthegithub commented:
I have improved the internals, moved the docs to typed, changes the docs, added a few ApiMayCahnged, moved part of the code in the internal package, added primitive tests for typed actors, changed the terminology from messages to events, and rebase onto the latest master.

patriknw commented:
Good thanks. Seems to be close to complete now, or do you have a list of missing pieces? I'll try to assign another reviewer to this.

manonthegithub committed:
api may change comment

manonthegithub commented:
> Good thanks. Seems to be close to complete now, or do you have a list of missing pieces?

@patriknw I planned to add a small comment in docs that it's a new api and may have breaking changes in the future, do you think it's needed? Another reviewer's comments would be nice. Persistent queries are missing, but its another PR.

raboof commented on akka-docs/src/main/paradox/typed/persistence-testing.md:
```diff
@@ -53,6 +53,121 @@ Java
+ * get all the events/snapshots persisted in the storage
+ * put the events/snapshots in the storage to test recovery
+ 
+<a id="setting-your-own-policy-for-the-storage"></a>
```
This explicit anchors is not necessary
```suggestion
```

manonthegithub replied:
removed

raboof commented on akka-docs/src/main/paradox/typed/persistence-testing.md:
```diff
@@ -53,6 +53,121 @@ Java
+#### Setting your own policy for the storage
+
+You can implement and set your own policy for the storage to control its actions on particular operations, for example you can fail or reject events on your own conditions.
+Implement the @scala[`ProcessingPolicy[EventStorage.JournalOperation] trait`]@java[`ProcessingPolicy<EventStorage.JournalOperation> interface`] for event storage
```
```suggestion
Implement the @apidoc[ProcessingPolicy[EventStorage.JournalOperation]] @scala[trait]@java[interface] for event storage
```

manonthegithub replied:
corrected

raboof commented on akka-docs/src/main/paradox/typed/persistence-testing.md:
```diff
@@ -53,6 +53,121 @@ Java
+
+You can implement and set your own policy for the storage to control its actions on particular operations, for example you can fail or reject events on your own conditions.
+Implement the @scala[`ProcessingPolicy[EventStorage.JournalOperation] trait`]@java[`ProcessingPolicy<EventStorage.JournalOperation> interface`] for event storage
+or @scala[`ProcessingPolicy[SnapshotStorage.SnapshotOperation] trait`]@java[`ProcessingPolicy<SnapshotStorage.SnapshotOperation> interface`] for snapshot storage,
```
```suggestion
or @apidoc[ProcessingPolicy[SnapshotStorage.SnapshotOperation]] @scala[trait]@java[interface] for snapshot storage,
```

manonthegithub replied:
corrected

raboof commented on akka-docs/src/main/paradox/typed/persistence-testing.md:
```diff
@@ -53,6 +53,121 @@ Java
+or @scala[`ProcessingPolicy[SnapshotStorage.SnapshotOperation] trait`]@java[`ProcessingPolicy<SnapshotStorage.SnapshotOperation> interface`] for snapshot storage,
+and set it with `withPolicy()` method.
+
+`tryProcess()` method of the `ProcessingPolicy` has two arguments: persistence id and the storage operation. 
```
```suggestion
`tryProcess()` method of the @apidoc[ProcessingPolicy] has two arguments: persistence id and the storage operation. 
```

manonthegithub replied:
correected

raboof reviewed (commented):
Good idea to move internal classes (such as `InMemStorage`) to `akka.persistence.testkit.internal`

raboof commented on akka-docs/src/main/paradox/typed/persistence-testing.md:
```diff
@@ -53,6 +53,121 @@ Java
+ * clear the events/snapshots persisted in the storage.
+ * reject the events, but not snapshots (rejections are not supported for snapshots in the original api).
+ * set your own [policy](#setting-your-own-policy-for-the-storage) which emulates the work of the storage. 
+Policy determines what to do when persistence needs to execute some operation on the storage (i.e. read, delete etc.).
```
```suggestion
Policy determines what to do when persistence needs to execute some operation on the storage (i.e. read, delete, etc.).
```

manonthegithub replied:
corrected

raboof commented on akka-persistence-testkit/src/test/scala/akka/persistence/testkit/scaladsl/PersistenceTestkitJournalCompatSpec.scala:
```diff
@@ -0,0 +1,42 @@
+        op match {
+          case WriteEvents(batch) =>
+            val allSerializable =
+              batch.filter(_.isInstanceOf[AnyRef]).forall(_.isInstanceOf[java.io.Serializable])
```
:+1: fair enough approximation :)

raboof commented on akka-persistence-testkit/src/test/scala/akka/persistence/testkit/javadsl/CommonTestkitTests.scala:
```diff
@@ -0,0 +1,575 @@
+import akka.util.ccompat.JavaConverters._
+import akka.actor.typed.javadsl.Adapter
+
+trait CommonTestkitTests extends JavaDslUtils {
```
More consistent to spell `TestKit` with a capital `K` in class/test names

manonthegithub replied:
corrected

raboof commented on akka-docs/src/test/java/jdocs/persistence/testkit/TestKitExamples.java:
```diff
@@ -0,0 +1,193 @@
+  // #set-event-storage-policy
+  class SampleEventStoragePolicy implements ProcessingPolicy<JournalOperation> {
+
+    // you can use internal state, it need not to be thread safe
```
```suggestion
    // you can use internal state, it does not need to be thread safe
```
(this text appears in some more places)

manonthegithub replied:
corrected

ignasi35 replied:
Hi @manonthegithub, I think the comment looks like the original (before @raboof's suggested change).

raboof commented on akka-persistence-testkit/src/test/scala/akka/persistence/testkit/javadsl/SnapshotNOTSerializeSpec.scala:
```diff
@@ -0,0 +1,26 @@
+import akka.actor.Props
+import akka.persistence.testkit._
+
+class SnapshotNOTSerializeSpec extends CommonSnapshotTests {
```
Maybe spell this `SnapshotNotSerializeSpec`?

manonthegithub replied:
corrected

raboof commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala:
```diff
@@ -0,0 +1,417 @@
+ * INTERNAL API
+ */
+@InternalApi
+private[testkit] class SerializedEventStorageImpl(system: ActorSystem) extends EventStorage {
```
This file is relatively large, perhaps move a couple of these implementation classes to their own files under an `internal` package?

manonthegithub replied:
made more splits and moves

raboof reviewed (commented):
Another review pass, mainly 'superficial' comments, in general looking good!

manonthegithub commented:
Thanks for the review, I'm currently very busy. I hope I will be able to finish it (according to the comments) in 1-3 weeks.

manonthegithub committed:
improvements according to the comments

manonthegithub committed:
improvements according to the comments + rebase on latest master and corresponding fixes

manonthegithub commented:
Rebased on the latest master and made changes according to @raboof's comments.

manonthegithub commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/ProcessingPolicy.scala:
```diff
@@ -0,0 +1,195 @@
+ * @tparam U type determines operations which storage can perform.
+ */
+@ApiMayChange
+trait ProcessingPolicy[U] {
```
I was reviewing the code again and thought that maybe it is worth to add a marker trait like `trait StorageOperation` and make the type parameter to depend on that `trait ProcessingPolicy[U <: StorageOperation]`, and then extend this trait with the JounalOperation or SnapshotOperation. It seems to me that it will limit the user, and specify that Policies work with storage operations. @patriknw @raboof do you think it is worth to do?

patriknw replied:
I have seen the question, but need to look a bit more into the policy structure before having an opinion.

patriknw replied:
I don't think that would be needed. Seems like it would add some more complexity to these trait hierarchies without much value.

When implementing a custom ProcessingPolicy it would be like:
```
class SampleEventStoragePolicy extends EventStorage.JournalPolicies.PolicyType
```

Maybe we could make it more concrete and easier to find by providing
```
trait JournalProcessingPolicy extends ProcessingPolicy[JournalOperation]
```
and the `withPolicy` could be defined as
```
def withPolicy(policy: JournalProcessingPolicy): this.type
```

Then it's more clear that a policy for Journal PersistenceTestKit is implemented by a JournalProcessingPolicy.
```
class SampleEventStoragePolicy extends JournalProcessingPolicy
```

similar for snapshots

manonthegithub replied:
I also had this idea :). Will try implement this. Last time I tried, there were some issues with it.

ignasi35 commented on akka-persistence-testkit/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,30 @@
+    assert-poll-interval = 100millis
+  }
+
+  # confugiration for persistence testkit for snapshots
```
```suggestion
  # configuiration for persistence testkit for snapshots
```

manonthegithub replied:
Nice catch! Fixed :)

manonthegithub committed:
fixed typos,
moved java serialization config parameter from conf file to the test

patriknw commented on akka-docs/src/test/java/jdocs/persistence/testkit/TestKitExamples.java:
```diff
@@ -0,0 +1,193 @@
+    @Override
+    public EventHandler<EmptyState, Evt> eventHandler() {
+      // some logic
+      return null;
```
suggestion instead of returning null:
```
      // TODO handle events
      return newEventHandlerBuilder()
        .build();
```

patriknw commented on akka-docs/src/test/java/jdocs/persistence/testkit/TestKitExamples.java:
```diff
@@ -0,0 +1,193 @@
+  }
+}
+
+class EmptyState {}
```
rename to `State`

also move State, Evt and Cmd into YourPersistentBehavior as `static class`

patriknw commented on akka-docs/src/test/scala/docs/persistence/testkit/TestKitExamples.scala:
```diff
@@ -0,0 +1,105 @@
+class TestKitExamples {
+
+  //#testkit-typed-usecase
+  class TypedSampleSpec extends WordSpecLike {
```
rename to `SampleSpec`

patriknw commented on akka-docs/src/test/scala/docs/persistence/testkit/TestKitExamples.scala:
```diff
@@ -0,0 +1,105 @@
+  //#testkit-typed-usecase
+  class TypedSampleSpec extends WordSpecLike {
+
+    val system: ActorSystem[Cmd] = ActorSystem(
```
better to use ScalaTestWithActorTestKit
https://doc.akka.io/docs/akka/current/typed/testing-async.html?language=scala#test-framework-integration

patriknw commented on akka-docs/src/test/scala/docs/persistence/testkit/TestKitExamples.scala:
```diff
@@ -0,0 +1,105 @@
+        emptyState = ???),
+      "name",
+      PersistenceTestKitPlugin.config.withFallback(ConfigFactory.defaultApplication()))
+    val testKit = PersistenceTestKit(system)
```
rename to avoid naming conflict with ScalaTestWithActorTestKit

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/PersistenceTestKitPlugin.scala:
```diff
@@ -0,0 +1,98 @@
+  private final val storage = InMemStorageExtension(context.system)
+
+  override def asyncWriteMessages(messages: immutable.Seq[AtomicWrite]): Future[immutable.Seq[Try[Unit]]] =
+    Future.fromTry(Try(messages.map(aw => storage.tryAdd(aw.payload))))
```
somewhere the `Tagged` should be unwrapped, so that asyncReplayMessages doesn't return the `Tagged`, as in the issue fixed in https://github.com/akka/akka/pull/28553

manonthegithub replied:
This is linked to #28602. I for now, until tck, and possibly some improvements to journal mechanics are made, I am planning just to ignore tags. Is that ok?

patriknw replied:
Would be nice if you unwrapped the `Tagged` before `tryAdd` here. In that way they are ignored, but it still works to use them for writing/replay.

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/Utils.scala:
```diff
@@ -0,0 +1,49 @@
+
+  private[testkit] object JavaCollectionConversions {
+
+    implicit def listConversion[A](jlist: JList[A]): immutable.Seq[A] =
```
I don't think we want the magical conversions between java and scala collections. Prefer the more explicit `.asScala` style as in `JavaConverters`.

Is this needed at all, or just use `JavaConverters`?

manonthegithub replied:
Wow. That's really good point. I think I used conversions, to avoid boilerplate of adding `.asJava`/`.asScala` in all the places.

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/Utils.scala:
```diff
@@ -0,0 +1,49 @@
+
+  }
+
+  private[testkit] object JavaFuncConversions {
```
similar here, not sure we should make it magic

patriknw reviewed (commented):
haven't reviewed all of it, but a few comments along the way
I can hopefully continue tomorrow.

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/SnapshotTestKit.scala:
```diff
@@ -0,0 +1,305 @@
+   * Set new processing policy for journal events.
+   * NOTE! Overrides previously invoked `failNext...` or `rejectNext...`
+   */
+  def withPolicy(policy: SnapshotStorage.SnapshotPolicies.PolicyType): this.type = {
```
`this.type` isn't great for Java. not sure if it will work at all, but this can be `: SnapshotTestKit` instead

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/scaladsl/TestOps.scala:
```diff
@@ -0,0 +1,350 @@
+            val ls = reprs.map(reprToAny)
+            assert(
+              ls.size == msgs.size && ls.zip(msgs).forall(e => e._1 == e._2),
+              "Persisted events do not correspond to expected ones")
```
include more info about the expected and actual events, otherwise difficult to see what's wrong

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/SnapshotTestKit.scala:
```diff
@@ -0,0 +1,305 @@
+    scalaTestkit.expectNothingPersisted(persistenceId, max.asScala)
+
+  /**
+   * Check that `msg` event has been saved in the storage.
```
rename `msg` to `event`

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/SnapshotTestKit.scala:
```diff
@@ -0,0 +1,305 @@
+  /**
+   * Check that next persisted in storage for particular persistence id event matches partial function `pf`.
+   */
+  def expectNextPersistedPF[A](persistenceId: String, pf: jf.Function[Any, A]): A =
```
Partial function for Java doesn't work well.

In Typed TestKit we have moved away from partial function asserts, also in scaladsl, because the error messages are not good. Instead we have the exact match with `expectMessage` and otherwise `expectMessageType` that can then be followed by more specific asserts using ScalaTest/JUnit.

Would be good to align this with Typed TestKit. See `akka.actor.testkit.typed.scaladsl.TestProbe` and `akka.actor.testkit.typed.javadsl.TestProbe`.

In other words, remove the PF methods and add `expectNextPersistedClass` in javadsl and `expectNextPersistedType` in scaladsl.

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/SnapshotTestKit.scala:
```diff
@@ -0,0 +1,305 @@
+  /**
+   * Check that `msgs` events have been persisted in the storage in order.
+   */
+  def expectPersistedInOrder[A](persistenceId: String, msgs: JList[A]): JList[A] =
```
a simplification for all these that assert several events would be to replace them with one single `receivePersisted(n: Int)` and the user can do the actual asserts using ScalaTest/JUnit.

See `receiveMessages` in `akka.actor.testkit.typed.scaladsl.TestProbe`

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/PersistenceTestKit.scala:
```diff
@@ -0,0 +1,478 @@
+  /**
+   * Check that next persisted in storage for particular persistence id event matches partial function `pf`.
+   */
+  def expectNextPersistedPF[A](persistenceId: String, pf: jf.Function[Any, A]): A =
```
see PF comment elsewhere

patriknw commented on akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/PersistenceTestKit.scala:
```diff
@@ -0,0 +1,478 @@
+  /**
+   * Check that `msgs` events have been persisted in the storage in order.
+   */
+  def expectPersistedInOrder[A](persistenceId: String, msgs: JList[A]): JList[A] =
```
see comment elsewhere

patriknw reviewed (commented):
Reviewed everything now. A few suggestions of simplifications but otherwise looking very good.

manonthegithub committed:
more improvements

manonthegithub commented:
Yep. Did exactly like that. I’m finishing improvements, just a few things left. I hope, I will push in a few days.

Kind regards / Mit freundlichen Grüßen,
Kirill Yankov.

> On 14 Feb 2020, at 04:04, Patrik Nordwall <notifications@github.com> wrote:
> 
> @patriknw commented on this pull request.
> 
> In akka-persistence-testkit/src/main/scala/akka/persistence/testkit/PersistenceTestKitPlugin.scala:
> 
> > +import scala.collection.immutable
> +import scala.concurrent.Future
> +import scala.util.Try
> +
> +/**
> + * INTERNAL API
> + *
> + * Persistence testkit plugin for events.
> + */
> +@InternalApi
> +class PersistenceTestKitPlugin extends AsyncWriteJournal {
> +
> +  private final val storage = InMemStorageExtension(context.system)
> +
> +  override def asyncWriteMessages(messages: immutable.Seq[AtomicWrite]): Future[immutable.Seq[Try[Unit]]] =
> +    Future.fromTry(Try(messages.map(aw => storage.tryAdd(aw.payload))))
> Would be nice if you unwrapped the Tagged before tryAdd here. In that way they are ignored, but it still works to use them for writing/replay.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

manonthegithub committed:
fix for Tagged events

manonthegithub commented:
Made corrections. Now should be fine.

manonthegithub committed:
formatting

manonthegithub committed:
fix docs examples compilation

manonthegithub commented:
Now it should be fine! Please check :)

ignasi35 commented on akka-docs/src/test/java/jdocs/persistence/testkit/TestKitExamples.java:
```diff
@@ -0,0 +1,180 @@
+    Evt expectedEventPersisted = new Evt(cmd.data);
+
+    persistenceTestKit.expectNextPersisted("your-persistence-id", expectedEventPersisted);
+    persistenceTestKit.clearAll();
```
The `pTK.clearAll()` only runs when the test completes successfully. This can cause a cascading failure of tests. Instead, we should suggest using `pTK.clearAll()` in a `@Before` method.

manonthegithub replied:
good catch!

ignasi35 commented on akka-docs/src/test/scala/docs/persistence/testkit/TestKitExamples.scala:
```diff
@@ -0,0 +1,105 @@
+        val expectedPersistedEvent = Evt(cmd.data)
+
+        persistenceTestKit.expectNextPersisted("your-persistence-id", expectedPersistedEvent)
+
```
The scala version of the `TestKitExamples.xxx` doesn`t showcase `pTK.clearAll` (as the java does). It'd be great to add it here too.

manonthegithub replied:
will add

ignasi35 reviewed (commented):
minor comments on the docs/examples.

manonthegithub commented:
@ignasi35 thanks for the review! I think the comment has been changed in scala version, but not at this particular line in java

manonthegithub committed:
- docs examples improvements
- scala 2.13 compilation fix

manonthegithub commented:
Corrected according to @ignasi35 's comments.

manonthegithub commented:
Just for planning my time. Do you guys have any plans for this?

patriknw commented:
I think we should merge this. It's ApiMayChange and also TestKit so we have full possibilities to iterate over things if we find things that should be adjusted. Let me take a final look tomorrow Friday.

patriknw reviewed (approved):
LGTM, very nice work @manonthegithub 
I'm going to merge this now. If others have more feedback it can be improved in new PRs. We have possibilities to adjust the APIs if needed.

manonthegithub commented:
Great! I think I will be able to support it for some time. Please ping me whenever there is an issue with it.

patriknw commented:
Thanks, if you want to add the query support that would be nice.

manonthegithub commented:
I will do the query support, but not sure about the timing. I hope I can do it in a month or two.

patriknw commented:
thanks
